Three level derived identity (against master)

mattdrees opened the pull request:
A test case and some work towards a solution for https://hibernate.atlassian.net/browse/HHH-5764

See related pull request #751

mattdrees commented:
I rebased against master and removed my non-test commits. I'll later take a crack at seeing if my 4.x changes will work against 5.x now. (Last year, they didn't.)

mattdrees commented:
So, looks like my changes for 4.x do indeed work against 5.x. Sweet.

I've added them here (after fixing conflicts).

(There is one failing test -- `GetAndIsVariantGetterTest.testAnnotations`, but I'm pretty sure it's not related to my changes)

sebersole commented:
Awesome, thanks Matt!  I will make this the last one I look at before I finally get some sleep tonight :)

BTW, GetAndIsVariantGetterTest passes currently on master and 5.0 branches.  So the problem would certainly seem to be in your changes.  I'll take a look when I pull this in unless you get to it first

sebersole commented on hibernate-core/src/main/java/org/hibernate/boot/internal/InFlightMetadataCollectorImpl.java:
```diff
@@ -1623,13 +1624,51 @@ private void processSecondPasses(ArrayList<? extends SecondPass> secondPasses) {
 			return;
 		}
 
-		for ( SecondPass secondPass : secondPasses ) {
```
Is there a reason to store these now-special SecondPasses in the generic SecondPass group?  If you notice one of the changes in 5.0 was to split these SecondPasses by type when they needed special handling or ordering.  Seems like this falls perfectly into that pattern.

sebersole commented on hibernate-core/src/main/java/org/hibernate/cfg/DependentSecondPass.java:
```diff
@@ -0,0 +1,9 @@
+ */
+public interface DependentSecondPass extends SecondPass {
+
+	boolean dependentUpon(SecondPass secondPass);
```
Coupled with my comment about storing these... I think this should accept `DependentSecondPass` arguments only.  And actually, since this only applies to `CopyIdentifierComponentSecondPass` I wonder if it might just make sense to use that for this contract.

I get that you are trying to make it generic.  But I'm not sure it gains us anything

mattdrees replied:
Thanks for the feedback!

Yeah, my sorting approach made more sense in the 4.x world than it does here. I left it mostly because I was lazy and wanted to get your input before fiddling more, but also partly because I think the sorting logic looks a little nicer using `DependentSecondPass` than `CopyIdentifierComponentSecondPass`.

I'll 'de-generify' this.

gbadner commented:
There are some really good tests in nestedcompositekeystest.zip attached to HHH-5764 using entities with @EmbeddedId and @IdClass that have fairly complex multi-level associations. When I ran them using Matt's pull request for 4.3, they all succeeded. I've moved them into hibernate-core unit tests and I'm ready to push once Matt has everything sorted out on master. Thanks Matt!

mattdrees commented:
@sebersole `GetAndIsVariantGetterTest.testAnnotations()` fails for me on 5.0 and master (even after a clean build). Not sure why. So I don't think that test failure is due to my changes. 

I'll try to get the other changes done here in the next day or two.

sebersole commented:
I'd like to understand the reason for that (maybe it is indicative of some
real platform/system specific bug).  Can you send me/us your JVM, OS, etc
info?


On Fri, Oct 30, 2015 at 12:45 PM Matt Drees <notifications@github.com>
wrote:

> @sebersole <https://github.com/sebersole>
> GetAndIsVariantGetterTest.testAnnotations() fails for me on 5.0 and
> master (even after a clean build). Not sure why. So I don't think that test
> failure is due to my changes.
>
> I'll try to get the other changes done here in the next day or two.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hibernate/hibernate-orm/pull/752#issuecomment-152598971>
> .
>

mattdrees commented:
from `groovy --version`:
Groovy:       2.3.10
Ant:          Apache Ant(TM) version 1.9.3 compiled on December 23 2013
JVM:          1.8.0_60 (Oracle Corporation 25.60-b23)
OS:           Mac OS X 10.10.5 x86_64

more details:
https://gist.github.com/mattdrees/0eb3b9243fa50a4b381b

mattdrees commented:
created a jira for you to track this: https://hibernate.atlassian.net/browse/HHH-10242

mattdrees commented:
I added a commit that removes the abstraction. 

However, I noticed that there's another bit of code in `InFlightMetadataCollectorImpl` that does a topological sort, as well: `processFkSecondPassesInOrder()`. I tried altering it to use the DependentSecondPass mechanism, and it seemed to work out well.  At the very least the tests pass. My work is located here: https://github.com/mattdrees/hibernate-orm/commit/5bbb7f7f37395d49481b10342e1cd4379e3c97b9. I'm guessing my sorting method is slower than the one I removed, but I think it's easier to understand.

Do you want to go with removing the abstraction? Or expanding its use?

sebersole commented:
Personally I am all for readability as long as the perf tradeoff is not
significant

On Mon, Nov 2, 2015, 5:46 PM Matt Drees <notifications@github.com> wrote:

> I added a commit that removes the abstraction.
>
> However, I noticed that there's another bit of code in
> InFlightMetadataCollectorImpl that does a topological sort, as well:
> processFkSecondPassesInOrder(). I tried altering it to use the
> DependentSecondPass mechanism, and it seemed to work out well. At the very
> least the tests pass. My work is located here: mattdrees@5bbb7f7
> <https://github.com/mattdrees/hibernate-orm/commit/5bbb7f7f37395d49481b10342e1cd4379e3c97b9>.
> I'm guessing my sorting method is slower than the one I removed, but I
> think it's easier to understand.
>
> Do you want to go with removing the abstraction? Or expanding its use?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hibernate/hibernate-orm/pull/752#issuecomment-153195080>
> .
>

mattdrees commented:
do you have a test somewhere that has hundreds (or thousands) of entities & relationships? I don't know of a good way to test the performance tradeoff.

sebersole commented:
You had mentioned the performance angle ;)

If we are unsure, then for now let's just keep that algorithm for `CopyIdentifierComponentSecondPass` processing.  We can look at `processFkSecondPassesInOrder` later

mattdrees commented:
Well, I really doubt it's a problem, but I just don't have a good way to verify that.

So yeah, let's just look at `processFkSecondPassesInOrder` later.


Anything else that you (or @gbadner ) want me to do here? squashing or whatnot?

sebersole commented:
Looks good to me.  Still waiting to hear from Gail wrt the question I asked
her on the Jira.  If she does not get to it or does not answer this week
I'll just apply it.

On Wed, Nov 4, 2015 at 6:35 PM Matt Drees <notifications@github.com> wrote:

> Well, I really doubt it's a problem, but I just don't have a good way to
> verify that.
>
> So yeah, let's just look at processFkSecondPassesInOrder later.
>
> Anything else that you (or @gbadner <https://github.com/gbadner> ) want
> me to do here? squashing or whatnot?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hibernate/hibernate-orm/pull/752#issuecomment-153915747>
> .
>

sebersole commented:
Applied upstream.  Thanks Matt!

mattdrees commented:
You're welcome! And thanks for applying this!
